internal/commands/build_test.go

```diff
@@ -709,6 +709,40 @@ builder = "my-builder"
 				h.AssertNil(t, command.Execute())
 			})
 		})
+
+		when("previous-image flag is provided", func() {
+			when("--publish is false", func() {
+				it("previous-image should be passed to creator", func() {
+					mockClient.EXPECT().
+						Build(gomock.Any(), EqBuildOptionsWithPreviousImage("previous-image")).
+						Return(nil)
+
+					command.SetArgs([]string{"--builder", "my-builder", "image", "--previous-image", "previous-image"})
+					h.AssertNil(t, command.Execute())
+				})
+			})
+
+			when("--publish is true", func() {
+				when("image and previous-image are in same registry", func() {
+					it("succeeds", func() {
+						mockClient.EXPECT().
+							Build(gomock.Any(), EqBuildOptionsWithPreviousImage("index.docker.io/some/previous:latest")).
+							Return(nil)
+
+						command.SetArgs([]string{"--builder", "my-builder", "index.docker.io/some/image:latest", "--previous-image", "index.docker.io/some/previous:latest", "--publish"})
+						h.AssertNil(t, command.Execute())
+					})
+				})
+
```

Comment:
Couldn't figure out this test case, get the following error message:
`=== RUN   TestBuildCommand/Commands/#BuildCommand/previous-image_flag_is_provided/--publish_is_true/image_and_previous-image_are_not_in_same_registry/fails
    build.go:130: Unexpected call to *testmocks.MockPackClient.Build([context.Background { index.docker.io/some/image:latest my-builder     map[] map[]  true false false [] [] <nil> { []}  always . {{ []} {[] [] [] [] } map[]}   -1 example.io/some/previous-image}]) at /Users/ujjwal/coding/pack/internal/commands/build.go:130 because: there are no expected calls of the method "Build" for that receiver`

Reply:
Mock assertions are always painful to read. In this case, it's complaining that your mock received a call, but your _test setup_ did not have the relevant `mockClient.EXPECT()` line.

Stepping back for a second. This unit test doesn't feel appropriate here. The logic for actually creating the error is way down in `internal/build/lifecycle_execution.go`. It feels like there's a decision to make about moving the test down or bringing the logic up...

I'd much prefer that you bring this logic (_and tests_) up. In fact, I'd prefer them to be at: **internal/commands/build.go**. That way we don't unnecessarily pull images, perform downloads, etc and then the action will fail anyway.

I'm glad that you tackled the edge case for us though 🙂

Reply:
Again, I added the test here because it was mentioned in the [issue](https://github.com/buildpacks/pack/issues/897#issuecomment-853059556). 😅 

I believe I can add the logic to `internal/commands/build.go`, but I don't get how to move the tests. Aren't the tests for this file in `internal/commands/build_test.go`, which is the file we're talking about here? Or would the logic of checking for invalid previous-image passed remove the need for a test? 😅 

cc: @jromero 

But for the sake of better understanding, I'd like to understand the error message.
> Mock assertions are always painful to read. In this case, it's complaining that your mock received a call, but your test setup did not have the relevant mockClient.EXPECT() line.

When I have the `mockClient` line, I got the error "expected error but got nil". I tried it with various registry names, but got the same result. For example, with previous image name `example.io/some/previous-image` and image name `index.docker.io/some/image:latest`. Apologies if this information is insufficient.

Also, thank you so much for helping me with this PR, really appreciate it 😄

Reply:
Just sent you a DM on slack if you want to do a bit of pair programing to get past this little hurdle.